### PR TITLE
VDS-145: Add customer segments domain models

### PR DESCRIPTION
**Pull request by asvishnyakov**
Also data models & CRUD operations logic added

### CODE REVIEW

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangedEvent.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+using System.Collections.Generic;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models;
+using VirtoCommerce.Platform.Core.Events;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Events
+{
+    public class DemoCustomerSegmentChangedEvent : GenericChangedEntryEvent<DemoCustomerSegment>
+    {
+        public DemoCustomerSegmentChangedEvent(IEnumerable<GenericChangedEntry<DemoCustomerSegment>> changedEntries) :
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangedEvent.cs]: Colon with calling base constructor together on new line.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangingEvent.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models;
+using VirtoCommerce.Platform.Core.Events;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Events
+{
+    public class DemoCustomerSegmentChangingEvent: GenericChangedEntryEvent<DemoCustomerSegment>
+    {
+        public DemoCustomerSegmentChangingEvent(IEnumerable<GenericChangedEntry<DemoCustomerSegment>> changedEntries) : base(changedEntries)
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangingEvent.cs]: :base(...

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegment.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+        public string Description { get; set; }
+
+        public bool IsActive { get; set; }
+
+        public DateTime? StartDate { get; set; }
+
+        public DateTime? EndDate { get; set; }
+
+        public ICollection<string> StoreIds { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegment.cs:24]: Initialize collection in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models
+{
+    public class DemoCustomerSegmentConditionEvaluationRequest
+    {
+        public IDictionary<string, string[]> Properties { get; set; }
+
+        public string Sort { get; set; }
+
+        public IList<SortInfo> SortInfos { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs:18]: Initialize collection in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models
+{
+    public class DemoCustomerSegmentConditionEvaluationRequest
+    {
+        public IDictionary<string, string[]> Properties { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs:14]: Initialize in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/Search/DemoCustomerSegmentSearchCriteria.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+using System.Collections.Generic;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Search
+{
+    public class DemoCustomerSegmentSearchCriteria: SearchCriteriaBase
+    {
+        public IList<string> StoreIds { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/Search/DemoCustomerSegmentSearchCriteria.cs:13]: Initialize collection in the constructor

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentConditionEvaluator.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Services
+{
+    public interface IDemoCustomerSegmentConditionEvaluator
+    {
+        public Task<string[]> EvaluateCustomerSegmentConditionAsync(DemoCustomerSegmentConditionEvaluationRequest conditionRequest);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentConditionEvaluator.cs]: "public" is redundant in interface

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentSearchService.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Search;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Services
+{
+    public interface IDemoCustomerSegmentSearchService
+    {
+        public Task<DemoCustomerSegmentSearchResult> SearchCustomerSegmentsAsync(DemoCustomerSegmentSearchCriteria criteria);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentSearchService.cs]: "public" is redundant

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+using System.Threading.Tasks;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Services
+{
+    public interface IDemoCustomerSegmentService
+    {
+        public Task<DemoCustomerSegment[]> GetByIdsAsync(string[] ids);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: public

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+        public Task SaveChangesAsync(DemoCustomerSegment[] customerSegments);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: public

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+        public Task DeleteAsync(string[] ids);
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: public

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/VirtoCommerce.DemoSolutionFeaturesModule.Core.csproj.DotSettings [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/VirtoCommerce.DemoSolutionFeaturesModule.Core.csproj.DotSettings:1]: Shouldn't we ignore this file?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    Id = table.Column<string>(maxLength: 128, nullable: false),
+                    CreatedDate = table.Column<DateTime>(nullable: false),
+                    ModifiedDate = table.Column<DateTime>(nullable: true),
+                    CreatedBy = table.Column<string>(maxLength: 64, nullable: true),
+                    ModifiedBy = table.Column<string>(maxLength: 64, nullable: true),
+                    Name = table.Column<string>(maxLength: 128, nullable: false),
+                    Description = table.Column<string>(maxLength: 1024, nullable: false),
+                    IsActive = table.Column<bool>(nullable: false),
+                    StartDate = table.Column<DateTime>(nullable: true),
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs:22]: StartDate is required, don't it?

Comment (pushnitsa) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs:22]: Shouldn't we add index to this field?

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/VirtoCommerce.DemoSolutionFeaturesModule.Core.csproj.DotSettings (hunk 12 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/VirtoCommerce.DemoSolutionFeaturesModule.Core.csproj.DotSettings:1]: No. Most of us use ReSharper and this file will say it don't treat some folders as namespace providers, i.e. 
`\src\VirtoCommerce.DemoSolutionFeaturesModule.Core\Models\Cart\DemoShoppingCart.cs`, but `VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.DemoShoppingCart`. It allowed me to move existing code into separate folder, but don't introduce breaking changes and has no warnings in ReSharper.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs (hunk 13 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs:22]: No, `StartDate` isn't required, we can leave it empty in blade. The only required field is StoreIds.
Why we need an index?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangedEvent.cs (hunk 1 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangedEvent.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangingEvent.cs (hunk 2 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Events/CustomerSegment/DemoCustomerSegmentChangingEvent.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegment.cs (hunk 3 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegment.cs:24]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs (hunk 4 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs:18]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs (hunk 5 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/DemoCustomerSegmentConditionEvaluationRequest.cs:14]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/Search/DemoCustomerSegmentSearchCriteria.cs (hunk 6 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/CustomerSegment/Search/DemoCustomerSegmentSearchCriteria.cs:13]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentConditionEvaluator.cs (hunk 7 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentConditionEvaluator.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentSearchService.cs (hunk 8 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentSearchService.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs (hunk 9 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs (hunk 10 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs (hunk 11 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Services/CustomerSegment/IDemoCustomerSegmentService.cs]: Fixed

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs (hunk 13 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Migrations/20201119110433_InitialCustomerSegments.cs:22]: After discussion: leave as is, ask why it set in FromModel in PromotionEntity

**Review by pushnitsa — DISMISSED** [review]

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**